### ISSUE TITLE
`Mem_Buf_Limit` + `Storage.type filesystem` seems not to work from v1.8.7

### ISSUE THREAD

**Issue by nokute78** [fluent/fluent-bit#4221]
## Bug Report

**Describe the bug**

In my understanding, `Mem_Buf_Limit` + `Storage.type filesystem` seems not to work from v1.8.7.

I think this patch affected.
https://github.com/fluent/fluent-bit/pull/4077/commits/01e791059de0262bea4f80d261b2f09352119089#diff-400dc531b019fa683ed5b361a69088bf0741d159d5408473a4801edbab5edae1L1106

**To Reproduce**
- Rubular link if applicable: None
- Example log message if applicable: Below

- Steps to reproduce the problem:
1. `mkdir tmp`
2. `cp a.conf tmp`
3. `sudo docker run -it --rm -v ${PWD}/tmp:/conf fluent/fluent-bit:1.8.7 /fluent-bit/bin/fluent-bit -c /conf/a.conf`

**Expected behavior**

**Screenshots**

**Your Environment**

* Version used: v1.8.7+
* Configuration:   a.conf:
```
[SERVICE]
    Storage.path hoge

[INPUT]
    Name dummy
    dummy {"msg":"hoge"}
    rate 1000
    Mem_Buf_Limit 1kb
    Storage.type filesystem
    Samples 1000

[OUTPUT]
    Name stdout
    Match *
```
* Environment name and version (e.g. Kubernetes? What version?): docker
* Server type and version: Ubuntu 20.04
* Operating System and version: Ubuntu 20.04
* Filters and plugins: none

**Additional context**

v1.8.6 outputs 1000 records (expected behavior).
v1.8.7 outputs only 48 records.

## v1.8.6

```
$ sudo docker run -it --rm -v ${PWD}/tmp:/conf fluent/fluent-bit:1.8.6 /fluent-bit/bin/fluent-bit -c /conf/a.conf
Fluent Bit v1.8.6
* Copyright (C) 2019-2021 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2021/10/24 09:00:05] [ info] [engine] started (pid=1)
[2021/10/24 09:00:05] [ info] [storage] created root path hoge
[2021/10/24 09:00:05] [ info] [storage] version=1.1.1, initializing...
[2021/10/24 09:00:05] [ info] [storage] root path 'hoge'
[2021/10/24 09:00:05] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2021/10/24 09:00:05] [ info] [storage] backlog input plugin: storage_backlog.1
[2021/10/24 09:00:05] [ info] [cmetrics] version=0.2.1
[2021/10/24 09:00:05] [ info] [input:storage_backlog:storage_backlog.1] queue memory limit: 95.4M
[2021/10/24 09:00:05] [ info] [sp] stream processor started
[0] hoge: [1635066005.780315352, {"msg"=>"hoge"}]
[1] hoge: [1635066005.781952138, {"msg"=>"hoge"}]
[2] hoge: [1635066005.782864437, {"msg"=>"hoge"}]
snip
[997] hoge: [1635066006.794237863, {"msg"=>"hoge"}]
[998] hoge: [1635066006.795047629, {"msg"=>"hoge"}]
[999] hoge: [1635066006.796229217, {"msg"=>"hoge"}]
^C[2021/10/24 09:00:11] [engine] caught signal (SIGINT)
[2021/10/24 09:00:11] [ info] [input] pausing storage_backlog.1
[2021/10/24 09:00:11] [ warn] [engine] service will stop in 5 seconds
[2021/10/24 09:00:16] [ info] [engine] service stopped

```

## v1.8.7

```
$ sudo docker run -it --rm -v ${PWD}/tmp:/conf fluent/fluent-bit:1.8.7 /fluent-bit/bin/fluent-bit -c /conf/a.conf
Fluent Bit v1.8.7
* Copyright (C) 2019-2021 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2021/10/24 08:59:16] [ info] [engine] started (pid=1)
[2021/10/24 08:59:16] [ info] [storage] created root path hoge
[2021/10/24 08:59:16] [ info] [storage] version=1.1.1, initializing...
[2021/10/24 08:59:16] [ info] [storage] root path 'hoge'
[2021/10/24 08:59:16] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2021/10/24 08:59:16] [ info] [storage] backlog input plugin: storage_backlog.1
[2021/10/24 08:59:16] [ info] [cmetrics] version=0.2.1
[2021/10/24 08:59:16] [ info] [input:storage_backlog:storage_backlog.1] queue memory limit: 95.4M
[2021/10/24 08:59:16] [ info] [sp] stream processor started
[2021/10/24 08:59:16] [ warn] [input] dummy.0 paused (mem buf overlimit)
[0] hoge: [1635065956.701288619, {"msg"=>"hoge"}]
(snip)
[47] hoge: [1635065956.749737584, {"msg"=>"hoge"}]
^C[2021/10/24 08:59:22] [engine] caught signal (SIGINT)
[2021/10/24 08:59:22] [ info] [input] pausing storage_backlog.1
[2021/10/24 08:59:22] [ warn] [engine] service will stop in 5 seconds
[2021/10/24 08:59:26] [ info] [engine] service stopped
```

**Comment by nokute78** [comment]
@pwhelan Any update ?

I think #4360 and #4376 is also same issue.

**Comment by edsiper** [comment]
@nokute78 can you try out this PR ?

https://github.com/fluent/fluent-bit/pull/4411

It worked locally, but I would like to get a second validation

**Comment by nokute78** [comment]
@edsiper Thank you for the patch.
I tested #4411 and it worked like v1.8.6. LGTM.

```
$ valgrind --leak-check=full ../bin/fluent-bit -c a.conf 
==4642== Memcheck, a memory error detector
==4642== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==4642== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==4642== Command: ../bin/fluent-bit -c a.conf
==4642== 
Fluent Bit v1.9.0
* Copyright (C) 2019-2021 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2021/12/09 19:30:39] [ info] [engine] started (pid=4642)
[2021/12/09 19:30:39] [ info] [storage] version=1.1.5, initializing...
[2021/12/09 19:30:39] [ info] [storage] root path 'hoge'
[2021/12/09 19:30:39] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2021/12/09 19:30:39] [ info] [storage] backlog input plugin: storage_backlog.1
[2021/12/09 19:30:39] [ info] [cmetrics] version=0.2.2
[2021/12/09 19:30:39] [ info] [input:storage_backlog:storage_backlog.1] queue memory limit: 95.4M
[2021/12/09 19:30:39] [ info] [sp] stream processor started
==4642== Warning: client switching stacks?  SP change: 0x57e59f8 --> 0x6c620f0
==4642==          to suppress, use: --max-stackframe=21481208 or greater
==4642== Warning: client switching stacks?  SP change: 0x6c62098 --> 0x57e59f8
==4642==          to suppress, use: --max-stackframe=21481120 or greater
==4642== Warning: client switching stacks?  SP change: 0x57e59f8 --> 0x6c62098
==4642==          to suppress, use: --max-stackframe=21481120 or greater
==4642==          further instances of this message will not be shown.
[0] dummy.0: [1639045839.800177180, {"msg"=>"hoge"}]
[1] dummy.0: [1639045839.866272270, {"msg"=>"hoge"}]
[2] dummy.0: [1639045839.868444414, {"msg"=>"hoge"}]
[3] dummy.0: [1639045839.868871366, {"msg"=>"hoge"}]
[4] dummy.0: [1639045839.869834406, {"msg"=>"hoge"}]
[5] dummy.0: [1639045839.870824253, {"msg"=>"hoge"}]
[6] dummy.0: [1639045839.871852626, {"msg"=>"hoge"}]
[7] dummy.0: [1639045839.872831715, {"msg"=>"hoge"}]
[8] dummy.0: [1639045839.873823755, {"msg"=>"hoge"}]
[9] dummy.0: [1639045839.874833617, {"msg"=>"hoge"}]
[10] dummy.0: [1639045839.875844690, {"msg"=>"hoge"}]
(snip)
[992] dummy.0: [1639045840.897896161, {"msg"=>"hoge"}]
[993] dummy.0: [1639045840.898888383, {"msg"=>"hoge"}]
[994] dummy.0: [1639045840.899891404, {"msg"=>"hoge"}]
[995] dummy.0: [1639045840.900934904, {"msg"=>"hoge"}]
[996] dummy.0: [1639045840.901915697, {"msg"=>"hoge"}]
[997] dummy.0: [1639045840.902896600, {"msg"=>"hoge"}]
[998] dummy.0: [1639045840.903898849, {"msg"=>"hoge"}]
[999] dummy.0: [1639045840.905132305, {"msg"=>"hoge"}]
^C[2021/12/09 19:31:10] [engine] caught signal (SIGINT)
[2021/12/09 19:31:10] [ info] [input] pausing storage_backlog.1
[2021/12/09 19:31:10] [ warn] [engine] service will shutdown in max 5 seconds
[2021/12/09 19:31:10] [ info] [engine] service has stopped (0 pending tasks)
==4642== 
==4642== HEAP SUMMARY:
==4642==     in use at exit: 0 bytes in 0 blocks
==4642==   total heap usage: 7,250 allocs, 7,250 frees, 26,059,878 bytes allocated
==4642== 
==4642== All heap blocks were freed -- no leaks are possible
==4642== 
==4642== For lists of detected and suppressed errors, rerun with: -s
==4642== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

**Comment by edsiper** [comment]
thanks @nokute78 !

**Comment by edsiper** [comment]
Fixed in #4411
